brainfuck/bf-syntax.rkt

```diff
@@ -0,0 +1,124 @@
+
+;;; Vector and tape ops.
+
+(define (vector-grow-if-needed vec len)
+  (define old-len (vector-length vec))
+  (cond [(< len old-len) vec]
+        [else
+         (let loop ([new-len (* 2 old-len)])
+           (cond [(>= len new-len) (loop (* 2 new-len))]
+                 [else (define new-vec (make-vector new-len))
+                       (vector-copy! new-vec 0 vec)
+                       new-vec]))]))
+
+(define (tape-get t)
+  (vector-ref (tape-data t) (tape-pos t)))
+
+(define (tape-move! t n)
+  (let ([new-pos (+ n (tape-pos t))])
+    (set-tape-data! t (vector-grow-if-needed (tape-data t) new-pos))
+    (set-tape-pos! t new-pos)))
+
+(define (tape-inc! t n)
+  (let ((data (tape-data t)) (pos (tape-pos t)))
+    (vector-set! data pos (+ n (vector-ref data pos)))))
+
+;;; Parser.
+
+(define (parse-helper lst acc)
+  (if (empty? lst)
+      (reverse acc)
+      (let ([rst (rest lst)])
+        (match (first lst)
+          [#\+ (parse-helper rst (cons (inc 1) acc))]
+          [#\- (parse-helper rst (cons (inc -1) acc))]
+          [#\> (parse-helper rst (cons (move 1) acc))]
+          [#\< (parse-helper rst (cons (move -1) acc))]
+          [#\. (parse-helper rst (cons -print acc))]
+          [#\[ (let ([subparsed (parse-helper rst empty)])
+                 (parse-helper (first subparsed)
+                               (cons (loop (rest subparsed)) acc)))]
+          [#\] (cons rst (reverse acc))]
+          [_ (parse-helper rst acc)]))))
+
+(define (parse bf-code) (parse-helper (string->list bf-code) empty))
+
+;;; Interpreter.
+
+(define ((inc x) t p) #`(tape-inc! #,t #,x))
+(define ((move x) t p) #`(tape-move! #,t #,x))
+(define (-print t p) #`(print #,p (tape-get #,t)))
+(define ((loop ops) t p)
+  #`(let go ()
+      (when (> (tape-get #,t) 0)
+        #,@(map (λ (op) (op t p)) ops)
+        (go))))
+
+(define (run parsed t p)
+  ((eval #`(λ (t p) #,@(map (λ (op) (op #'t #'p)) parsed))) t p))
+
+(define (notify msg)
+  (with-handlers ([exn:fail:network? (lambda (_) (void))])
+    (let-values ([(in out) (tcp-connect "localhost" 9001)])
+      (display msg out)
+        (close-output-port out))))
+
+(define (read-c path)
+  (parameterize ([current-locale "C"])
+    (file->string path)))
+
+(define (verify)
+  (define text "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>\
+---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.>++.")
+  (define p-left (printer 0 0 #t))
+  (define p-right (printer 0 0 #t))
+
+  (run (parse text) (tape (vector 0) 0) p-left)
+  (for-each
+   (lambda (c) (print p-right (char->integer c)))
+   (string->list "Hello World!\n"))
+
+  (let ((left (get-checksum p-left))
+        (right (get-checksum p-right)))
+    (when (not (eq? left right))
+        (error 'verify "~s != ~s" left right))))
+
+(module+ main
+  (verify)
+  (define text (read-c (command-line #:args (filename) filename)))
+  (define p (printer 0 0 (getenv "QUIET")))
+
+  (notify (format "Racket\t~s" (getpid)))
```

Comment:
Could you please change the test name (as the current one would conflict with the existing one), something like "Racket (Syntax Objects)"? I'm not sure how to call it though, I guess you know better.

Reply:
I changed it to `Racket (Syntax Objects)`. I would normally just call it something like `Racket (Compile)`, but as you can see that easily misleads that it's cheating ^_^

Reply:
Also, I just realized that this is faster in Racket 8.0. So it would be great if you can update that some time!

Reply:
Yes, Racket 8.0 is used in the up-coming maintenance update.